src/core/ddsc/src/dds_entity.c

```diff
@@ -1007,6 +1012,12 @@ static void pushdown_listener (dds_entity *e)
 
       ddsrt_mutex_lock (&e->m_observers_lock);
       dds_override_inherited_listener (&c->m_listener, &e->m_listener);
+      if (ddsrt_avl_is_empty(&c->m_children)) {
```

Comment:
What you are basically saying here is that since the events only occur on the ultimate children, it doesn't hurt to let for example a Participant or a Publisher check its OfferedIncompatibleQosStatus bit, since that bit will never be set on that level? That might be true, and the check might actually prevent a pending DATA_ON_READERS to be invoked, because a subscriber probably does have children..... 
I will remove this check in my pull-request.